File: cpp/src/global_planning/search/jps.cpp
#include "navigation/global_planning/search/jps.hpp"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdlib>
#include <optional>
#include <queue>
#include <unordered_map>
#include <utility>
#include <vector>

#include "navigation/global_planning/search/discrete_search.hpp"  // emit_finished_discrete

namespace navigation::global_planning {

using core::DynamicGridSpace;

namespace {

const double kSqrt2 = std::sqrt(2.0);

// Octile distance on integer cell deltas, in the same operation order —
// (hi - lo) + sqrt(2)*lo — as OccupancyGrid2D::heuristic. A pure straight/diagonal
// jump line has this as its true traversed cost, so JPS f-values and the returned
// cost match 8-connected A* exactly. Admissible.
double octile(const Cell& a, const Cell& b) {
  int dr = std::abs(a.row - b.row);
  int dc = std::abs(a.col - b.col);
  int lo = std::min(dr, dc);
  int hi = std::max(dr, dc);
  return static_cast<double>(hi - lo) + kSqrt2 * static_cast<double>(lo);
}

int sgn(int x) { return (x > 0) - (x < 0); }

// Single-cell occupancy oracle over DynamicGridSpace::is_blocked (occupied OR out of
// bounds) — the only grid primitive JPS reads.
struct Oracle {
  DynamicGridSpace<Cell>& space;
  bool blocked(int r, int c) const { return space.is_blocked(Cell{r, c}); }
  bool freec(int r, int c) const { return !space.is_blocked(Cell{r, c}); }
};

// First jump point along (dr,dc) from (r,c), or nullopt if the run dead-ends at an
// obstacle / boundary. Recursion depth is at most 1: a diagonal run's orthogonal
// probes are straight scans that never recurse (Harabor & Grastien 2011).
std::optional<Cell> scan(const Oracle& o, int r, int c, int dr, int dc, const Cell& goal) {
  const bool diagonal = dr != 0 && dc != 0;
  while (true) {
    // Step legality mirrors the map's corner rule: a diagonal needs the target and
    // both shared orthogonal cells free (no corner cutting).
    if (o.blocked(r + dr, c + dc)) return std::nullopt;
    if (diagonal && (o.blocked(r + dr, c) || o.blocked(r, c + dc))) return std::nullopt;
    r += dr;
    c += dc;
    if (r == goal.row && c == goal.col) return Cell{r, c};
    if (!diagonal) {
      if (dc != 0) {  // horizontal: obstacle diagonally behind opens a side cell
        if ((o.freec(r - 1, c) && o.blocked(r - 1, c - dc)) ||
            (o.freec(r + 1, c) && o.blocked(r + 1, c - dc)))
          return Cell{r, c};
      } else {  // vertical
        if ((o.freec(r, c - 1) && o.blocked(r - dr, c - 1)) ||
            (o.freec(r, c + 1) && o.blocked(r - dr, c + 1)))
          return Cell{r, c};
      }
    } else if (scan(o, r, c, dr, 0, goal).has_value() ||
               scan(o, r, c, 0, dc, goal).has_value()) {
      // A diagonal cell whose orthogonal scan finds a jump point is itself one.
      return Cell{r, c};
    }
  }
}

// Directions to jump from u given how it was reached. Start (no parent) explores all
// eight; otherwise natural continuation + forced-neighbour branches. Illegal branches
// simply yield no jump point, so a slightly liberal set never costs correctness.
std::vector<std::pair<int, int>> successor_dirs(const Oracle& o, const Cell& u,
                                                const std::optional<Cell>& parent) {
  static const std::pair<int, int> kAll[8] = {{-1, 0}, {1, 0},  {0, -1}, {0, 1},
                                              {-1, -1}, {-1, 1}, {1, -1}, {1, 1}};
  // Parenthesised (iterator-range) ctor — a braced list here would try to build an
  // initializer_list<pair> from two pointers.
  if (!parent.has_value()) return std::vector<std::pair<int, int>>(kAll, kAll + 8);
  const int pdr = sgn(u.row - parent->row);
  const int pdc = sgn(u.col - parent->col);
  if (pdr != 0 && pdc != 0) {  // diagonal: continuation + both orthogonal legs
    return {{pdr, 0}, {0, pdc}, {pdr, pdc}};
  }
  std::vector<std::pair<int, int>> dirs{{pdr, pdc}};  // natural continuation
  if (pdc != 0) {                                     // horizontal
    if (o.freec(u.row - 1, u.col) && o.blocked(u.row - 1, u.col - pdc)) {
      dirs.push_back({-1, pdc});
      dirs.push_back({-1, 0});
    }
    if (o.freec(u.row + 1, u.col) && o.blocked(u.row + 1, u.col - pdc)) {
      dirs.push_back({1, pdc});
      dirs.push_back({1, 0});
    }
  } else {  // vertical
    if (o.freec(u.row, u.col - 1) && o.blocked(u.row - pdr, u.col - 1)) {
      dirs.push_back({pdr, -1});
      dirs.push_back({0, -1});
    }
    if (o.freec(u.row, u.col + 1) && o.blocked(u.row - pdr, u.col + 1)) {
      dirs.push_back({pdr, 1});
      dirs.push_back({0, 1});
    }
  }
  return dirs;
}

// JPS's parent chain is sparse (jump points only). Fill the straight/diagonal cells
// between consecutive jump points so the reported path is the full staircase, matching
// how A* reports every cell.
std::vector<Cell> reconstruct_full(const std::unordered_map<Cell, Cell>& parent,
                                   const Cell& start, const Cell& goal) {
  std::vector<Cell> jumps;
  Cell cur = goal;
  jumps.push_back(cur);
  while (!(cur == start)) {
    cur = parent.at(cur);
    jumps.push_back(cur);
  }
  std::reverse(jumps.begin(), jumps.end());
  std::vector<Cell> path{start};
  for (size_t i = 1; i < jumps.size(); ++i) {
    const Cell a = jumps[i - 1];
    const Cell b = jumps[i];
    const int dr = sgn(b.row - a.row);
    const int dc = sgn(b.col - a.col);
    Cell c = a;
    while (!(c == b)) {
      c = Cell{c.row + dr, c.col + dc};
      path.push_back(c);
    }
  }
  return path;
}

}  // namespace

core::PlanResult<Cell> JpsPlanner::plan(DynamicGridSpace<Cell>& space, const Cell& start,
                                        const Cell& goal, TraceRecorder* recorder) {
  Oracle oracle{space};
  struct QItem {
    double f;
    unsigned long long seq;  // insertion order; FIFO tie-break keeps ties stable
    Cell cell;
  };
  struct Greater {
    bool operator()(const QItem& a, const QItem& b) const {
      // Same (f, seq) tie-break as A* so the deterministic search is stable.
      return a.f > b.f || (a.f == b.f && a.seq > b.seq);
    }
  };

  auto t0 = std::chrono::steady_clock::now();
  std::priority_queue<QItem, std::vector<QItem>, Greater> open;
  std::unordered_map<Cell, double> g;
  std::unordered_map<Cell, Cell> parent;
  std::unordered_map<Cell, bool> settled;
  unsigned long long seq = 0;

  g[start] = 0.0;
  open.push({octile(start, goal), seq++, start});

  PlanResult<Cell> result;
  PlanStats stats;
  bool found = false;

  while (!open.empty()) {
    Cell u = open.top().cell;
    open.pop();
    if (settled[u]) continue;  // stale duplicate with a worse key
    settled[u] = true;
    if (recorder) recorder->node_expanded(u, g[u]);
    ++stats.expanded_nodes;
    if (u == goal) {
      found = true;
      break;
    }
    std::optional<Cell> par;
    if (auto it = parent.find(u); it != parent.end()) par = it->second;
    for (const auto& [dr, dc] : successor_dirs(oracle, u, par)) {
      std::optional<Cell> jp = scan(oracle, u.row, u.col, dr, dc, goal);
      if (!jp.has_value() || settled[*jp]) continue;
      double ng = g[u] + octile(u, *jp);
      auto it = g.find(*jp);
      if (it == g.end() || ng < it->second) {
        g[*jp] = ng;
        parent[*jp] = u;
        if (recorder) recorder->candidate_evaluated(*jp, ng);
        if (recorder) recorder->edge_added(*jp, u, octile(u, *jp));
        open.push({ng + octile(*jp, goal), seq++, *jp});
      }
    }
  }

  if (found) {
    result.success = true;
    result.path = reconstruct_full(parent, start, goal);
    result.cost = g[goal];  // jump-point path cost; NOT path_cost() (sums per-edge only).
    if (recorder) recorder->path_found(result.path);
  }
  result.stats = stats;
  double rt = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();
  emit_finished_discrete(recorder, result.success, result.cost, stats, rt);
  return result;
}

}  // namespace navigation::global_planning

File: python/navigation/global_planning/search/jps.py
"""Jump Point Search (JPS) — A* over grid "jump points".

Harabor & Grastien (2011). On a uniform-cost 8-connected grid, most of A*'s
expansions are wasted on symmetric permutations of the same path. JPS breaks that
symmetry: instead of every neighbour, a node's successors are the *jump points*
reached by scanning each canonical direction until an obstacle / the goal / a cell
that has a *forced neighbour* (a neighbour reachable optimally only through the
current cell because of an adjacent obstacle). It returns exactly the same optimal
paths as 8-connected A* while expanding far fewer nodes.

This repository's grid forbids corner-cutting (a diagonal step needs both shared
orthogonal cells free — see ``OccupancyGrid2D.neighbors``), so the pruning uses the
no-corner-cutting variant of the rules in Harabor & Grastien (2011): a straight run
has a forced neighbour when an obstacle sits diagonally *behind* the travel
direction, opening a side cell that can only be reached by turning at the current
cell; a diagonal run cuts no corners, so its side openings are already covered by
its two orthogonal scans and it needs no separate forced neighbour.

Capabilities: JPS is a discrete 8-connected grid search (``DISCRETE_SPACE``) whose
jump primitive is a single-cell occupancy/bounds oracle — exactly
``DynamicGridSpace.is_blocked`` (occupied OR out of bounds), so it binds to
``DynamicGridSpace`` and reads no ground-truth neighbour list. Octile costs are
computed here (like Theta* computes its own Euclidean distance) because the bound
space exposes no ``heuristic``; the formula matches ``OccupancyGrid2D.heuristic`` so
the returned cost equals the 8-connected A* optimum bit-for-bit.
"""

from __future__ import annotations

import heapq
import itertools
import math
import time

from navigation.core.capabilities import Capability, DynamicGridSpace
from navigation.core.planner import GlobalPlanner
from navigation.core.trace import TraceRecorder
from navigation.core.types import Cell, PlanResult, PlanStats

from ._discrete import emit_finish

_SQRT2 = math.sqrt(2.0)
# The eight canonical directions, orthogonals before diagonals — the same order as
# OccupancyGrid2D.neighbors, so the start node (which explores all directions)
# settles ties identically to A*.
_DIRS_8 = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)]


def _octile(a: Cell, b: Cell) -> float:
    # Octile distance on integer cell deltas, in exactly the same operation order —
    # (hi - lo) + sqrt(2)*lo, sqrt NOT hypot — as OccupancyGrid2D.heuristic. A pure
    # straight/diagonal jump line has this as its true traversed cost, so successor
    # edge costs and f-values match 8-connected A* exactly. Admissible.
    dr = abs(a[0] - b[0])
    dc = abs(a[1] - b[1])
    lo = min(dr, dc)
    hi = max(dr, dc)
    return float(hi - lo) + _SQRT2 * float(lo)


def _sign(x: int) -> int:
    return (x > 0) - (x < 0)


class JPS(GlobalPlanner[Cell, "DynamicGridSpace[Cell]"]):
    @property
    def name(self) -> str:
        return "jps"

    def required_capabilities(self) -> set[Capability]:
        # DISCRETE_SPACE: JPS is 8-connected grid graph search (same optimal paths as
        # A*). DYNAMIC_GRID_SPACE: is_blocked is the single-cell oracle the jumps need.
        return {Capability.DISCRETE_SPACE, Capability.DYNAMIC_GRID_SPACE}

    def plan(
        self,
        space: DynamicGridSpace[Cell],
        start: Cell,
        goal: Cell,
        recorder: TraceRecorder | None = None,
    ) -> PlanResult[Cell]:
        t0 = time.monotonic()

        def blocked(r: int, c: int) -> bool:
            return space.is_blocked((r, c))

        def free(r: int, c: int) -> bool:
            return not space.is_blocked((r, c))

        def scan(r: int, c: int, dr: int, dc: int) -> Cell | None:
            # Advance from (r,c) along (dr,dc) one legal grid step at a time; return the
            # first jump point (goal, or a cell with a forced neighbour) or None if the
            # run dead-ends at an obstacle / boundary. Recursion depth is at most 1: a
            # diagonal run's orthogonal probes are straight scans that never recurse.
            diagonal = dr != 0 and dc != 0
            while True:
                # Step legality mirrors the map's corner rule: a diagonal needs the
                # target and both shared orthogonal cells free (Harabor & Grastien 2011,
                # no-corner-cutting variant).
                if blocked(r + dr, c + dc):
                    return None
                if diagonal and (blocked(r + dr, c) or blocked(r, c + dc)):
                    return None
                r, c = r + dr, c + dc
                if (r, c) == goal:
                    return (r, c)
                if not diagonal:
                    if dc != 0:  # horizontal: obstacle diagonally behind opens a side cell
                        if (free(r - 1, c) and blocked(r - 1, c - dc)) or (
                            free(r + 1, c) and blocked(r + 1, c - dc)
                        ):
                            return (r, c)
                    else:  # vertical
                        if (free(r, c - 1) and blocked(r - dr, c - 1)) or (
                            free(r, c + 1) and blocked(r - dr, c + 1)
                        ):
                            return (r, c)
                elif (
                    scan(r, c, dr, 0) is not None or scan(r, c, 0, dc) is not None
                ):
                    # A diagonal cell whose orthogonal scan finds a jump point is itself a
                    # jump point: the optimal path must branch here.
                    return (r, c)

        def successor_dirs(u: Cell, parent: Cell | None) -> list[tuple[int, int]]:
            # Directions to jump from u given how it was reached. Start (no parent)
            # explores all eight. Otherwise: natural continuation + forced-neighbour
            # branches (Harabor & Grastien 2011). Illegal branches simply yield no jump
            # point, so a slightly liberal set never costs correctness.
            if parent is None:
                return _DIRS_8
            r, c = u
            pdr, pdc = _sign(r - parent[0]), _sign(c - parent[1])
            if pdr != 0 and pdc != 0:  # diagonal: continuation + both orthogonal legs
                return [(pdr, 0), (0, pdc), (pdr, pdc)]
            dirs: list[tuple[int, int]] = [(pdr, pdc)]  # natural continuation
            if pdc != 0:  # horizontal
                if free(r - 1, c) and blocked(r - 1, c - pdc):
                    dirs += [(-1, pdc), (-1, 0)]
                if free(r + 1, c) and blocked(r + 1, c - pdc):
                    dirs += [(1, pdc), (1, 0)]
            else:  # vertical
                if free(r, c - 1) and blocked(r - pdr, c - 1):
                    dirs += [(pdr, -1), (0, -1)]
                if free(r, c + 1) and blocked(r - pdr, c + 1):
                    dirs += [(pdr, 1), (0, 1)]
            return dirs

        g: dict[Cell, float] = {start: 0.0}
        parent: dict[Cell, Cell] = {}
        closed: set[Cell] = set()
        counter = itertools.count()  # stable tie-break, keeps heap entries comparable
        frontier: list[tuple[float, int, Cell]] = [(_octile(start, goal), next(counter), start)]
        expanded = 0
        found = False
        while frontier:
            _, _, u = heapq.heappop(frontier)
            if u in closed:
                continue
            closed.add(u)
            expanded += 1
            if recorder is not None:
                recorder.node_expanded(u, g[u])
            if u == goal:
                found = True
                break
            for dr, dc in successor_dirs(u, parent.get(u)):
                jp = scan(u[0], u[1], dr, dc)
                if jp is None or jp in closed:
                    continue
                tentative = g[u] + _octile(u, jp)
                if jp not in g or tentative < g[jp]:
                    g[jp] = tentative
                    parent[jp] = u
                    if recorder is not None:
                        recorder.candidate_evaluated(jp, tentative)
                        recorder.edge_added(jp, u, _octile(u, jp))
                    heapq.heappush(frontier, (tentative + _octile(jp, goal), next(counter), jp))

        runtime = time.monotonic() - t0
        if not found:
            emit_finish(recorder, False, [], 0.0, expanded, runtime)
            return PlanResult(success=False, stats=PlanStats(expanded_nodes=expanded))
        path = _reconstruct_full(parent, start, goal)
        cost = g[goal]
        emit_finish(recorder, True, path, cost, expanded, runtime)
        return PlanResult(True, path, cost, PlanStats(expanded_nodes=expanded))


def _reconstruct_full(parent: dict[Cell, Cell], start: Cell, goal: Cell) -> list[Cell]:
    # JPS's parent chain is sparse (jump points only). Fill the straight/diagonal grid
    # cells between consecutive jump points so the reported path is the full staircase,
    # matching how A* reports every cell (viz + path-connectivity checks rely on it).
    jumps = [goal]
    node = goal
    while node != start:
        node = parent[node]
        jumps.append(node)
    jumps.reverse()
    path = [start]
    for a, b in zip(jumps, jumps[1:], strict=False):
        dr, dc = _sign(b[0] - a[0]), _sign(b[1] - a[1])
        cur = a
        while cur != b:
            cur = (cur[0] + dr, cur[1] + dc)
            path.append(cur)
    return path
